retrieval/aggregator.go

```diff
@@ -0,0 +1,159 @@
+
+// CounterAggregatorMetricConfig provides configuration of a single aggregated counter.
+// Matchers specify what Prometheus metrics (which are expected to be counter metrics) will
+// be re-aggregated. Help provides a description for the exported metric.
+type CounterAggregatorMetricConfig struct {
+	Matchers [][]*promlabels.Matcher
+	Help     string
+}
+
+// counterTracker keeps track of a single time series that has at least one aggregated
+// counter associated with it (i.e. there is at least one aggregated counter that has
+// Matchers covering this time series). Last timestamp and value are tracked
+// to detect counter resets.
+type counterTracker struct {
+	lastTimestamp int64
+	lastValue     float64
+	measures      []*stats.Float64Measure
+	logger        log.Logger
+}
+
+// NewCounterAggregator creates a counter aggregator.
+func NewCounterAggregator(logger log.Logger, config *CounterAggregatorConfig) (*CounterAggregator, error) {
+	aggregator := &CounterAggregator{logger: logger}
+	for metric, cfg := range *config {
+		measure := stats.Float64(metric, cfg.Help, stats.UnitDimensionless)
+		v := &view.View{
+			Name:        metric,
+			Description: cfg.Help,
+			Measure:     measure,
+			Aggregation: view.Sum(),
+		}
+		if err := view.Register(v); err != nil {
+			return nil, err
+		}
+		aggregator.counters = append(aggregator.counters, &aggregatedCounter{
+			measure:  measure,
+			view:     v,
+			matchers: cfg.Matchers,
+		})
+	}
+	return aggregator, nil
+}
+
+// Close must be called when CounterAggregator is no longer needed.
+func (c *CounterAggregator) Close() {
+	for _, counter := range c.counters {
+		view.Unregister(counter.view)
+	}
+}
+
+// getTracker returns a counterTracker for a specific time series defined by labelset.
+// If `nil` is returned, it means that there are no aggregated counters that need to
+// be incremented for this time series.
+func (c *CounterAggregator) getTracker(lset labels.Labels) *counterTracker {
+	var measures []*stats.Float64Measure
+	for _, counter := range c.counters {
+		if matchFiltersets(lset, counter.matchers) {
+			measures = append(measures, counter.measure)
+		}
+	}
+	if len(measures) == 0 {
+		return nil
+	}
+	return &counterTracker{measures: measures, logger: c.logger}
+}
+
+// newPoint gets called on each new sample (timestamp, value) for time series that need to feed
+// values into aggregated counters.
+func (a *counterTracker) newPoint(ctx context.Context, lset labels.Labels, t int64, v float64) {
+	if math.IsNaN(v) {
+		level.Debug(a.logger).Log("msg", "got NaN value", "labels", lset, "last ts", a.lastTimestamp, "ts", t, "lastValue", a.lastValue)
+		return
+	}
+	// Ignore points that are earlier than last seen timestamp.
+	if t < a.lastTimestamp {
+		level.Debug(a.logger).Log("msg", "out of order timestamp", "labels", lset, "last ts", a.lastTimestamp, "ts", t)
+		return
+	}
+	// First time we're seeing a value; record it, but don't increment counters.
```

Comment:
Minor. How about, "Use the first value we see as the baseline for this counter"? It reads more like the intention and less like the literal code.

Reply:
Done.